h263_codec_registration.src.html

```diff
@@ -0,0 +1,154 @@
+==============================================================
+
+The codec string begins with the prefix `"h263."`, with a suffix of 1 character
+describing the version of the specification: `"h263.1"` for the first revision
+([[H.263v1]]), `"h263.2"` for the second revision ([[H.263v2]]), or `"h263.3"`
+for the third revision ([[H.263v3]]). As the specifications are backwards
+compatible, an encoder may encode using an earlier revision than specified, and
+a decoder may decode using a later revision than specified.
+
+`"h263.3"`, but no prior revision, may further be suffixed with a `"."` and a
+one-character profile number, as specified in Annex X.2 of [[H.263v3]]. If a
+profile is not specified, then an encoder may use any optional features and a
+decoder must support all optional features. This may further be suffixed by a
+`"."` and a two-character level, as described in Table X.2 of [[H.263v3]]. If a
+level is not specified, then an encoder may support any level or may support
+formats other than those in any specified level, and a decoder must support any
+level and may support formats other than those in any specified level.
+
+As revisions of H.263 prior to H.263v3 do not have profiles or levels, if
+`"h263.1"` or `"h263.2"` is specified, then an encoder may support any optional
+features in the specified revision, and a decoder must support all optional
+features in the specified revision. The non-normative profiles and levels
+specified in Appendix II are not used.
+
+EncodedVideoChunk data {#encodedvideochunk-data}
+================================================
+
+{{EncodedVideoChunk}} [=EncodedVideoChunk/[[internal data]]=] is expected to be
+a picture as described in Section 5.1 of [[H.263v1]], [[H.263v2]], or
+[[H.263v3]], as appropriate for the used specification revision.
+
+VideoDecoderConfig description {#videodecoderconfig-description}
+================================================================
+
+{{VideoDecoderConfig.description}} is not used for this codec.
+
+EncodedVideoChunk type {#encodedvideochunk-type}
+================================================
+
+If an {{EncodedVideoChunk}}'s [=EncodedVideoChunk/[[type]]=] is
+{{EncodedVideoChunkType/key}}, then the {{EncodedVideoChunk}} is expected to
+contain an `INTRA` picture. The precise definition depends on the specification
+revision:
+
+H.263v1: `PTYPE`'s bit 9 (Picture Coding Type) is 0 (`INTRA`), as defined in
+Section 5.1.3 of [[H.263v1]].
+
+H.263v2, H.263v3: If `PTYPE`'s bits 6-8 are not set to 111, then `PTYPE`'s bit
+9 (Picture Coding Type) is 0 (`INTRA`), as defined in Section 5.1.3. If
+`PTYPE`'s bits 6-8 are set to 111, as defined in Section 5.1.3, then
+`PLUSPTYPE`'s `MPPTYPE`'s bits 1-3 (Picture Type Code) are 000 (`INTRA`), as
+defined in Section 5.1.4.3. Section numbers refer to [[H.263v2]] if the second
+revision is used, or [[H.263v3]] if the third revision is used.
+
+Notes on codec naming {#notes-on-codec-naming}
```

Comment:
I typically see codec strings break down by profiles rather than annexes. I appreciate that profiles came along pretty late here, but I wonder if they could still serve this purpose and thereby simplify the codec string scheme. At a glance, it looks like the profiles refer to combinations of features from earlier annexes. Would it work to simply have a string like h263.[profile]?

Reply:
The problem is that the profiles and levels don't have a clean relationship with the revisions, but encoders and decoders tend to be revision-specific. Only profile 0 is compatible with the original H.263v1, but it doesn't use all the features H.263v1 is allowed to use. Profiles 1–3 are compatible with H.263v2, but leave out features it's allowed to use. Any practical implementation is going to use an off-the-shelf en/decoder, and ffmpeg for instance has h263 (v1) and h263p (v2) as their own codecs.

Reply:
The problem is that the profiles don't have a clean relationship with the revisions, but practical codecs are by revision. Profile 0 is only decodable with H.263v1, but doesn't use all of H.263v1's features. Profiles 1–3 are decodable with H.263v2, but don't use all of _its_ features. Profiles 4–8 all require H.263v3. Any practical implementation is going to use off-the-shelf codecs, and in practice, the codecs are by revision. For instance, ffmpeg has "h263" (v1) and "h263p" (v2).